### PR TITLE
Add tile template option to raster and vector sources

### LINKED ISSUE #6245: MGLRasterSource and MGLVectorSource tile sets
MGLRasterSource needs support for the `tiles`, `minzoom`, and `maxzoom` style specification properties. We’ll need to add `tileURLTemplates`, `minimumZoomLevel`, and `maximumZoomLevel` properties to MGLRaster(Vector)Source, which we’ll use to create an `mbgl::style::TileSet` to pass into the `mbgl::style::RasterSource` constructor.

/cc @frederoni @boundsj

### PR DESCRIPTION
**Pull request by boundsj**
- [x] introduce MGLTileSet
- [x] add tileset initializer to MGLVectorSource
- [x] test MGLVectorSource in iosapp
- [x] add tileset initializer to MGLRasterSource 
- [x] test MGLRasterSource in iosapp
- [x] unit tests
- [x] documentation updates

Fixes #6245

### CODE REVIEW

File: platform/darwin/src/MGLRasterSource.h [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
+#import "MGLTypes.h"
+
+NS_ASSUME_NONNULL_BEGIN
 
 @interface MGLRasterSource : MGLSource
 
 @property (nonatomic, readonly, copy) NSURL *URL;
 @property (nonatomic, readonly, assign) CGFloat tileSize;
+@property (nonatomic, readonly, copy) NS_ARRAY_OF(NSString *) *tileURLTemplates;
+@property (nonatomic, readonly) uint8_t minimumZoomLevel;
```
Comment (1ec5) [inline platform/darwin/src/MGLRasterSource.h]: These should be `NSInteger`-typed. In general, we should prefer Foundation typealiases over the compiler’s built-in numeric primitives in Objective-C code. It’s fine if the whole range of `NSInteger` won’t work; if it fails hard, we can clamp the value to whatever range is supported.

File: platform/darwin/src/MGLRasterSource.h [hunk 2]
```diff
@@ -1,10 +1,19 @@
 #import "MGLSource.h"
+#import "MGLTypes.h"
+
+NS_ASSUME_NONNULL_BEGIN
 
 @interface MGLRasterSource : MGLSource
 
 @property (nonatomic, readonly, copy) NSURL *URL;
 @property (nonatomic, readonly, assign) CGFloat tileSize;
```
Comment (1ec5) [inline platform/darwin/src/MGLRasterSource.h]: `-mbglSource` casts this value to `uint16_t`, so this property shouldn’t be a floating-point number. (Or is the internal tile size in pixels rather than points?)

File: platform/darwin/src/MGLRasterSource.mm [hunk 3]
```diff
@@ -6,18 +6,48 @@
...
+}
+
+- (std::unique_ptr<mbgl::style::Source>)mbglSource
+{
+    std::unique_ptr<mbgl::style::RasterSource> source;
+    
+    if (self.URL)
+    {
+        source = std::make_unique<mbgl::style::RasterSource>(self.sourceIdentifier.UTF8String,
+                                                             self.URL.absoluteString.UTF8String,
```
Comment (1ec5) [inline platform/darwin/src/MGLRasterSource.mm]: Call `-[NSURL(MGLAdditions) mgl_URLByStandardizingScheme]` in order to canonicalize asset: URLs and such.

File: platform/darwin/src/MGLRasterSource.mm [hunk 4]
```diff
@@ -6,18 +6,48 @@
...
+                                                             uint16_t(self.tileSize));
+    }
+    else
+    {
+        mbgl::Tileset tileSource;
+        tileSource.zoomRange = mbgl::Range<uint8_t>(self.minimumZoomLevel, self.maximumZoomLevel);
+        for (NSString *tileURLTemplate in self.tileURLTemplates)
+        {
+            tileSource.tiles.push_back(tileURLTemplate.UTF8String);
```
Comment (1ec5) [inline platform/darwin/src/MGLRasterSource.mm]: Do we need something similar to `-mgl_URLByStandardizingScheme` for URL template strings? I guess it’d be pretty impractical to store tiles locally and include them via a URL template.

**Comment by 1ec5** [comment]
We should probably add an MGLTileSet class that exposes all the options `mbgl::style::TileSet` does and accept that in the MGLRasterSource initializer instead of the URL templates. Otherwise, it wouldn’t be possible to set any of the other TileJSON properties such as `data` or `attribution` (which will become important once #5999 lands).

**Comment by boundsj** [comment]
> We should probably add an MGLTileSet class

This work has started in c0fb939

File: platform/darwin/src/MGLTileSet.mm [hunk 5]
```diff
@@ -0,0 +1,72 @@
+#import "MGLTileSet.h"
+
+#include <mbgl/util/tileset.hpp>
+
+@implementation MGLTileSet
+{
+    BOOL _minimumZoomLevelIsSet;
```
Comment (1ec5) [inline platform/darwin/src/MGLTileSet.mm]: If the minimum and maximum zoom levels are optional, then do they need to be represented as NSNumbers publicly? Otherwise, if you use the `-initWithTileURLTemplates:` initializer, these properties end up being 0.

**Comment by boundsj** [comment]
With the latest changes here, you can now mash up vector and raster sources created with tile sets:

```ObjC
    // Use raster tile URL templates directly (i.e. mapbox://tiles/mapbox.satellite/{z}/{x}/{y}{ratio}.png)
    // This is Stamen's terrain
    MGLTileSet *rasterTileSet = [[MGLTileSet alloc] initWithTileURLTemplates:@[@"http://a.tile.stamen.com/terrain-background/{z}/{x}/{y}.jpg"]];
    MGLRasterSource *rasterSource = [[MGLRasterSource alloc] initWithSourceIdentifier:@"raster-source" tileSize:256 tileSet:rasterTileSet];
    [self.mapView.style addSource:rasterSource];
    
    MGLRasterStyleLayer *rasterLayer = [[MGLRasterStyleLayer alloc] initWithLayerIdentifier:@"raster-layer" source:rasterSource];
    [self.mapView.style addLayer:rasterLayer];
    
    // Use vector tile URL templates directly. This is Mapzen's vector tile service
    MGLTileSet *vectorTileSet = [[MGLTileSet alloc] initWithTileURLTemplates:@[@"https://vector.mapzen.com/osm/all/{z}/{x}/{y}.mvt?api_key=akey"]];
    MGLVectorSource *vectorSource = [[MGLVectorSource alloc] initWithSourceIdentifier:@"mapzen" tileSet:vectorTileSet];
    [self.mapView.style addSource:vectorSource];
    
    MGLFillStyleLayer *buildingsFillLayer = [[MGLFillStyleLayer alloc] initWithLayerIdentifier:@"filllayer" source:vectorSource sourceLayer:@"buildings"];
    buildingsFillLayer.fillColor = [UIColor magentaColor];
    [self.mapView.style addLayer:buildingsFillLayer];
```

#### Not exactly "Streets"
screen shot 2016-09-21 at 3 00 31 pm

The current implementation does not attempt to reflect the mbgl object's default minimum and maximum zoom values if the Objective-C object's properties are nil. However, documentation was added that states that the defaults will be used if the Objective-C tile set object is used to initialize a source and the min/max values are nil.

cc @1ec5

File: platform/darwin/src/MGLTileSet.h [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ The `MGLTileSet` class holds the tile URL template strings and associated
+ configuration for those strings. It can be passed to a `MGLVectorSource` or
+ `MGLRasterSource` instead of an `NSURL` representing a TileJSON URL to create a 
+ source.
+ */
+@interface MGLTileSet : NSObject
+
+/**
+ A REQUIRED `NSArray` of `NSString` objects that represent the tile templates.
```
Comment (1ec5) [inline platform/darwin/src/MGLTileSet.h]: Optionality should be indicated through nullability qualifiers; shouting isn't needed here.

File: platform/darwin/src/MGLTileSet.h [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+ 
+ @param tileURLTemplates An `NSArray` of `NSString` objects that represent the 
+ tile templates.
+ @param minimumZoomLevel An `NSNumber` object containing an integer; specifies 
+ the minimum zoom level at which to display tiles.
+ @param maximumZoomLevel An `NSNumber` object containing an integer; specifies
+ the maximum zoom level at which to display tiles.
+ @return The initialized tile set object.
+ */
+- (instancetype)initWithTileURLTemplates:(NS_ARRAY_OF(NSString *) *)tileURLTemplates minimumZoomLevel:(NSNumber *)minimumZoomLevel maximumZoomLevel:(NSNumber *)maximumZoomLevel;
```
Comment (1ec5) [inline platform/darwin/src/MGLTileSet.h]: The zoom level parameters should either be nullable or, better yet, primitive typed.

File: platform/ios/test/MGLTileSetTests.mm [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    // a tile set that provides an mbgl tile set
+    MGLTileSet *tileSet = [[MGLTileSet alloc] initWithTileURLTemplates:@[@"tile.1",
+                                                                         @"tile.2",
+                                                                         @"tile.3"]];
+    mbgl::Tileset mbglTileset = [tileSet mbglTileset];
+    
+    // has the correct URL templates
+    XCTAssertEqual(mbglTileset.tiles[0], "tile.1");
+    XCTAssertEqual(mbglTileset.tiles[1], "tile.2");
+    XCTAssertEqual(mbglTileset.tiles[2], "tile.3");
```
Comment (1ec5) [inline platform/ios/test/MGLTileSetTests.mm]: Also assert on the tile set count.

File: platform/ios/test/MGLTileSetTests.mm [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    // when the tile set has min and/or max zoom level set
+    tileSet.minimumZoomLevel = @(1);
+    tileSet.maximumZoomLevel = @(2);
+    
+    // the mbgl object reflects the set values for min and max zoom level
+    XCTAssertEqual([tileSet mbglTileset].zoomRange.min, 1);
+    XCTAssertEqual([tileSet mbglTileset].zoomRange.max, 2);
+    
+    // when the tile set has an attribution
+    tileSet.attribution = @"my tileset";
```
Comment (1ec5) [inline platform/ios/test/MGLTileSetTests.mm]: Also test that Unicode characters are preserved.

File: platform/ios/test/MGLTileSetTests.mm [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    tileSet.minimumZoomLevel = nil;
+    tileSet.maximumZoomLevel = nil;
+    
+    // the mbgl object has default values for min and max zoom level
+    XCTAssertEqual([tileSet mbglTileset].zoomRange.min, 0);
+    XCTAssertEqual([tileSet mbglTileset].zoomRange.max, 22);
+    
+    // when the tile set has min and/or max zoom level set
+    tileSet.minimumZoomLevel = @(1);
+    tileSet.maximumZoomLevel = @(2);
```
Comment (1ec5) [inline platform/ios/test/MGLTileSetTests.mm]: If the developer sets a maximum zoom level less than the minimum, what happens? Would it be possible to test for an Objective-C exception being thrown (ideal)?

Comment (boundsj) [inline platform/ios/test/MGLTileSetTests.mm]: 0fc9cbc throws Objective-C exceptions if values are crossed initially or later on in setters. This avoids a more shocking C++ exception that happens later when the doomed threshold is crossed when the map is zoomed. 

Alternatively, we could ignore invalid values and log a warning avoiding an exception altogether.

File: platform/darwin/src/MGLRasterSource.h (hunk 2 above)
Comment (boundsj) [inline platform/darwin/src/MGLRasterSource.h]: f7a4427 changes this to an `NSUInteger`. I think the tile size here represents "tile units" as discussed in https://github.com/mapbox/mapbox-gl-native/issues/6236

File: platform/darwin/src/MGLTileSet.h [hunk 11]
```diff
@@ -11,43 +11,43 @@ typedef NS_ENUM(NSUInteger, MGLTileSetScheme) {
...
 @property (nonatomic, copy) NS_ARRAY_OF(NSString *) *tileURLTemplates;
 
 /**
- An OPTIONAL `NSNumber` object containing an integer; specifies the minimum zoom
- level at which the source will display tiles. The value should be in the range 
- of 0 to 22. The default value is 0 and the source will use the default value
+ An `NSNumber` object containing an integer; specifies the minimum zoom level at 
+ which the source will display tiles. The value should be in the range of 0 to 
+ 22. The default value is 0 and the source will use the default value
  if `minimumZoomLevel` is nil;
```
Comment (1ec5) [inline platform/darwin/src/MGLTileSet.h]: This sentence and two more below end in semicolons instead of periods.

File: platform/darwin/src/MGLTileSet.h [hunk 12]
```diff
@@ -11,43 +11,43 @@ typedef NS_ENUM(NSUInteger, MGLTileSetScheme) {
...
+ which to display tiles. The value should be in the range of 0 to 22 and greater 
+ than `minimumZoomLevel`. The default value is 22 and the source will use the 
+ default value if `maximumZoomLevel` is nil;
  */
 @property (nonatomic, nullable) NSNumber *maximumZoomLevel;
 
 /**
- An OPTIONAL `NSString` object; Contains an attribution to be displayed
- when the map is shown to a user. The default value is nil;
+ An `NSString` object; Contains an attribution to be displayed when the map is
```
Comment (1ec5) [inline platform/darwin/src/MGLTileSet.h]: Nit: replace the semicolon with “that” and lowercase “contains” (plus one more below).

File: platform/ios/test/MGLTileSetTests.mm [hunk 13]
```diff
@@ -43,7 +44,7 @@ - (void)testMBGLTileSet {
     XCTAssertEqual([tileSet mbglTileset].zoomRange.max, 2);
     
     // when the tile set has an attribution
-    tileSet.attribution = @"my tileset";
+    tileSet.attribution = @"my tileset ©";
```
Comment (1ec5) [inline platform/ios/test/MGLTileSetTests.mm]: © is a decent test, since it falls outside of ASCII. In #5999, I used ©️, which pairs © with a variation selector for an emoji look. But something outside the Basic Multilingual Plane, for example in the [Miscellaneous Symbols and Pictographs](https://en.wikipedia.org/wiki/Miscellaneous_Symbols_and_Pictographs) block, would be even better for detecting mojibake.

Comment (boundsj) [inline platform/ios/test/MGLTileSetTests.mm]: Thanks! I think 0e31030 should probably wrap this up
